module/zfs/spa_errlog.c

```diff
@@ -404,6 +900,134 @@ spa_errlog_sync(spa_t *spa, uint64_t txg)
 	mutex_exit(&spa->spa_errlog_lock);
 }
 
+static void
+delete_dataset_errlog(spa_t *spa, uint64_t spa_err_obj, uint64_t ds,
+    dmu_tx_t *tx)
+{
+	if (spa_err_obj == 0)
+		return;
+
+	zap_cursor_t zc;
+	zap_attribute_t za;
+	for (zap_cursor_init(&zc, spa->spa_meta_objset, spa_err_obj);
+	    zap_cursor_retrieve(&zc, &za) == 0; zap_cursor_advance(&zc)) {
+		uint64_t head_ds;
+		name_to_object(za.za_name, &head_ds);
+		if (head_ds == ds) {
+			(void) zap_remove(spa->spa_meta_objset, spa_err_obj,
+			    za.za_name, tx);
+			VERIFY(dmu_object_free(spa->spa_meta_objset,
+			    za.za_first_integer, tx) == 0);
+			break;
+		}
+	}
+	zap_cursor_fini(&zc);
+}
+
+void
+spa_delete_dataset_errlog(spa_t *spa, uint64_t ds, dmu_tx_t *tx)
+{
+	mutex_enter(&spa->spa_errlog_lock);
+	delete_dataset_errlog(spa, spa->spa_errlog_scrub, ds, tx);
+	delete_dataset_errlog(spa, spa->spa_errlog_last, ds, tx);
+	mutex_exit(&spa->spa_errlog_lock);
+}
+
+#if defined(_KERNEL)
```

Comment:
Is `#if defined(_KERNEL)` needed here?

Reply:
`swap_errlog()` references `name_to_errphys()` which is also in an `ifdef _KERNEL` above.

Reply:
I could move `name_to_errphys()` above out of its `ifdef _KERNEL` block, and that would allow removing those ifdefs.